src/store/Controller.cc

```diff
@@ -354,7 +338,7 @@ Store::Controller::find(const cache_key *key)
                 allowSharing(*entry, key);
             checkFoundCandidate(*entry);
             entry->touch();
-            referenceBusy(*entry);
+            removeFromReplacementPurgePolicies(*entry);
```

Comment:
The old code refreshed policy lists for disk and memory (placing the found entry at the tail of the list by calling *_eferenced() methods. Now we (still) have one 'purge' list for memory and two lists, 'purge' and 'walk'. The new code does this refreshing only for 'purge' lists (actually this happens when we re-add the entry that becomes idle). I think that there should be no benefit in doing this for 'walk' disk list, that is used for just traversing all objects in statObjects() and storeCleanup().

Reply:
> meant to end the above sentence with "... for each Store::Disk",

Yes, for each disk.

> statObjects() uses Store::Root().search().
> storeCleanup() uses Store::Root().search()
> StoreSearchUFS is unused.

I missed the fact that StoreSearchPointer in these methods is never StoreSearchUFS and also that StoreSearchUFS is unused itself.

> The only place where the new "walk" list is (or should be) used in the PR code is ...  storeDirWriteCleanLogs()

There is one more place where this list is used - Store::Disk::stat(). However, this method does not iterate entries but calls a single 'stat' method, e.g., lru_stats().

>  no idle ones in rock environments.

It looks like that rock does not create its policy Disk::repl field at all (see Rock::SwapDir::parse()) and its methods SwapDir::trackReferences()/SwapDir::ignoreReferences() have no effect.

>  I am right about StoreSearchUFS being unused,

Agreed.

> study pros and cons to find the best fix

From the memory consumption point of view, storing one small and one big lists ('idle' and 'busy') is better that storing two big lists ('idle' and 'all'). We could even optimize by moving nodes between idle<->busy instead of allocating new nodes. On the other hand, without a single 'all' list we would not be able to use it in Store::Disk::stat(), as it was used before. However, replWalk->Stats() does little, filling only ' reference age' - and probably we could take it from 'busy' list instead (fortunately these lru/heap stats() methods do not return e.g., the list size, so we do not need to merge 'busy' and 'idle' lists in this case).

> writing that list either before the "idle" list or "after" 

I see no sense in writing the 'busy' list before the 'idle' list if we are going to preserve the elements order. In this order elements are purged from the head of the lists, and fresh/recently_accessed elements are at the tail.

Reply:
> post an official PR to remove it 

Posted an official PR1285.

Reply:
> There is one more place where this list is used - Store::Disk::stat(). However, this method does not iterate entries but calls a single 'stat' method, e.g., lru_stats().

> without a single 'all' list we would not be able to use it in Store::Disk::stat(), as it was used before. However, replWalk->Stats() does little, filling only ' reference age' - and probably we could take it from 'busy' list instead

AFAICT, after looking at this statistics collection code: lru_stat() is the only entry list-related statistics collection functionality used by Store::Disk::stat(). lru_stat() reports the reference age of the very first unlocked StoreEntry on the LRU policy list (i.e. the least recently used unlocked entry). If we have an LRU list of idle entries, the first entry on that list will be used to report exactly the same information without a goto-powered linear search. This can be viewed as a yet another optimization and code improvement introduced by this PR.

Reply:
> converting the "walk" list into a "busy"

Converted.